sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala

```diff
@@ -314,6 +329,78 @@ object GpuShuffledHashJoinExec {
     }
   }
 
+  private def getBuildIterMaybeFiltered(
+      gpuBuildIter: Iterator[ColumnarBatch],
+      buildGoal: CoalesceSizeGoal): (Iterator[ColumnarBatch], Boolean) = {
+    var accBatchRows = 0L
+    closeOnExcept(new mutable.ArrayBuffer[SpillableColumnarBatch]) { spillBuf =>
+      // Check if the total rows number goes beyond the hard limit of 2^31 rows
+      while (accBatchRows < Int.MaxValue && gpuBuildIter.hasNext) {
+        closeOnExcept(gpuBuildIter.next()) { cb =>
+          accBatchRows += cb.numRows()
+          spillBuf.append(
+            SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_BATCHING_PRIORITY))
+        }
+      }
+      val rowsExceedIntMax = accBatchRows > Int.MaxValue ||
+        (accBatchRows == Int.MaxValue && gpuBuildIter.hasNext)
+      val safeIter = GpuSubPartitionHashJoin.safeIteratorFromSeq(spillBuf).map { sp =>
+        withResource(sp)(_.getColumnarBatch())
+      } ++ gpuBuildIter
+
+      buildGoal match {
+        case RequireSingleBatchWithFilter(filterExpr) if rowsExceedIntMax =>
+          val filteredIter = safeIter.map(cb => withResource(cb)(GpuFilter(_, filterExpr)))
+          (filteredIter, true)
+        case _ => (safeIter, false)
+      }
+    }
+  }
+
+  private def getFilteredBuildBatches(
+      filteredIter: Iterator[ColumnarBatch],
+      targetSize: Long,
+      subPartConf: Option[Boolean],
+      buildTime: GpuMetric): Either[ColumnarBatch, Iterator[ColumnarBatch]] = {
+    // Redo the size estimation for filtered batches
+    var accBatchSize = 0L
+    closeOnExcept(new mutable.ArrayBuffer[SpillableColumnarBatch]) { spillBuf =>
+      while (accBatchSize < targetSize && filteredIter.hasNext) {
+        closeOnExcept(filteredIter.next()) { cb =>
+          accBatchSize += GpuColumnVector.getTotalDeviceMemoryUsed(cb)
+          spillBuf.append(
+            SpillableColumnarBatch(cb, SpillPriorities.ACTIVE_BATCHING_PRIORITY))
+        }
+      }
+      val multiBuildBatches = subPartConf.getOrElse {
+        // Total size goes beyond the target size.
+        accBatchSize > targetSize || (accBatchSize == targetSize && filteredIter.hasNext)
+      }
+      if (multiBuildBatches) {
+        // The size still overflows after filtering or sub-partitioning is enabled for test.
+        logWarning("Return multiple batches as the build side data for the following " +
+          "sub-partitioning join in null-filtering mode.")
+        val safeIter = GpuSubPartitionHashJoin.safeIteratorFromSeq(spillBuf).map { sp =>
+          withResource(sp)(_.getColumnarBatch())
+        } ++ filteredIter
+        Right(new CollectTimeIterator("hash join build", safeIter, buildTime))
+      } else { // We can get here only when the filtered size is within the target size.
+        while(filteredIter.hasNext) {
```

Comment:
Shouldn't this be an assert that there isn't any more to read?  If we are here it means we looked at all of the data and it is small enough.

Reply:
updated

Reply:
I'v changed it back for the case when sub-partitioning is disabled by setting the relevant conf to false.